In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\dbnewman18\Instacart_Basket_Analysis_31_03-2022'

In [33]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_test = pd.DataFrame()

In [6]:
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prods.shape

(49693, 5)

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

In [24]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Task 4.5

In [26]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max price of $99,999 is too high for a grocery product.

In [27]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [35]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No columns with mixed data types found.

In [36]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [37]:
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [38]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [39]:
df_order_check=df_ords_nan[df_ords_nan['order_number']!= 1]

In [40]:
df_order_check

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

The days_since_prior_order values are only missing for rows in which the order number is equal to 1. This means that in these cases there is no prior order because it is the customer's first order.

In [42]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [44]:
df_ords_nan_check=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [45]:
df_ords_nan_check

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [46]:
df_ords_zero_check=df_ords[df_ords['order_number']==1]

In [47]:
df_ords_zero_check

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     0.0  
11                        2                 11                     0.0  
26                        1                 14                     0.0  
39                        6                 11                     0.0  
45                        3                 12                     0.0  
...                     ...                ...                     ...  
3420930                   4                 12                     0.0  
3420934                   3                 18                     0.0  
3421002                   6                 19                     0.0  
3421019                   1                 15                     0.0  
3421069                   3                 11                     0.0  

[206209 rows x 8 columns]

In [48]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [49]:
df_ords['days_since_prior_order'].dtype

dtype('float64')

In [50]:
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].astype('int64')

In [52]:
df_ords_zero_check.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
11          11   2168274        2    prior             1                   2   
26          26   1374495        3    prior             1                   1   
39          39   3343014        4    prior             1                   6   
45          45   2717275        5    prior             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     0.0  
11                 11                     0.0  
26                 14                     0.0  
39                 11                     0.0  
45                 12                     0.0

In [54]:
df_ords['days_since_prior_order'].dtype

dtype('int64')

Dealt with the missing values by replacing the NaNs with zeroes. Not sure why the decimal point remains even after changing data type to integer. Maybe you could explain this?

In [55]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [56]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [57]:
df_ords[df_ords.duplicated('order_id')]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [58]:
df_ords[df_ords.duplicated('user_id')]

Unnamed: 0  order_id  user_id eval_set  order_number  \
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
5                 5   3367565        1    prior             6   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
1                         3                  7                      15  
2                         3                 12                      21  
3                         4                  7                      29  
4                         4                 15                      28  
5                         2                  7                      19  
...                     ...                ...                     ...  
3421078                   5                 18                      29  
3421079                   4                 10                      30  
3421080                   1                 12                      18  
3421081                   1                 12                       7  
3421082                   6                 14                      30  

[3214874 rows x 8 columns]

These duplicates exist when the same user has more than one order. They can be left as is.

In [59]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orrders_checked.csv'))